c/uca/sifter/ucd.hpp

```diff
@@ -0,0 +1,244 @@
+// © 2024 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+#pragma once
+#include <algorithm>
+#include <array>
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+#include <map>
+#include <ranges>
+#include <string>
+#include <vector>
+
+namespace unicode {
+
+class CheckFailure {
+   public:
+    ~CheckFailure() {
+        std::terminate();
+    }
+    void operator&(std::ostream&) {}
+};
+
+#define CHECK(condition) \
+    (condition)          \
+            ? (void)0    \
+            : CheckFailure() & std::cerr << ("Check failed: " #condition "\n")
+
+class CodePointRange {
+   public:
+    struct iterator {
+        constexpr char32_t operator*() const {
+            return value;
+        }
+
+        constexpr void operator++() {
+            ++value;
+        }
+
+        constexpr std::strong_ordering operator<=>(
+                const iterator& other) const = default;
+
+        char32_t value;
+    };
+
+    static constexpr CodePointRange inclusive(char32_t first, char32_t last) {
+        return CodePointRange(first, last + 1);
+    }
+
+    constexpr bool empty() const {
+        return first_ >= pastTheEnd_;
+    }
+
+    constexpr std::size_t size() const {
+        return empty() ? 0 : pastTheEnd_ - first_;
+    }
+
+    constexpr bool contains(char32_t c) const {
+        return c >= first_ && c < pastTheEnd_;
+    }
+
+    constexpr CodePointRange intersection(const CodePointRange& other) const {
+        return CodePointRange(std::max(first_, other.first_),
+                              std::min(pastTheEnd_, other.pastTheEnd_));
+    }
+
+    constexpr char32_t front() const {
+        CHECK(!empty());
+        return first_;
+    }
+
+    constexpr char32_t back() const {
+        CHECK(!empty());
+        return pastTheEnd_ - 1;
+    }
+
+    iterator begin() const {
+        return {first_};
+    }
+
+    iterator end() const {
+        return {pastTheEnd_};
+    }
+
+   private:
+    constexpr CodePointRange(char32_t first, char32_t past_the_end)
+        : first_(first), pastTheEnd_(past_the_end) {}
+
+    char32_t first_;
+    char32_t pastTheEnd_;
+};
+
+// TODO(egg): Make this behave like a container.
+class CodePointSet {
+   public:
```

Comment:
An indentation of three looks very weird. I prefer labels without indentation. Indenting by half of the normal might also be ok.
```suggestion
public:
```

Reply:
Ditto, will look at doing that in .clang-format.